scripts/update-browser-releases/bun.ts

```diff
@@ -0,0 +1,347 @@
+ * @param dir The directory to search within.
+ * @returns The path to the Bun binary, or null if not found.
+ */
+const findBunBinaryPath = async (dir: string): Promise<string | null> => {
+  const entries = await fs.readdir(dir, { withFileTypes: true });
+  for (const e of entries) {
+    const p = path.join(dir, e.name);
+    if (e.isDirectory()) {
+      const found = await findBunBinaryPath(p);
+      if (found) {
+        return found;
+      }
+    } else if (e.isFile() && e.name === 'bun') {
+      return p;
+    }
+  }
+  return null;
+};
+
+/**
+ * Downloads and extracts a specific Bun version for the current platform.
+ * Uses a local cache and reuses artifacts when available.
+ * @param version The Bun version to download, for example "1.2.21".
+ * @returns The absolute path to the extracted Bun binary, or null on failure.
+ */
+const downloadAndExtractBun = async (
+  version: string,
+): Promise<string | null> => {
+  const suffix = resolveAssetSuffix();
+  const base = `https://github.com/oven-sh/bun/releases/download/bun-v${version}`;
+  const asset = `bun-${suffix}.zip`;
+  const url = `${base}/${asset}`;
+
+  const cacheRoot = path.join(process.cwd(), '.cache', 'bun-updater', version);
+  const zipPath = path.join(cacheRoot, asset);
+  const extractDir = path.join(cacheRoot, 'extracted');
+
+  await fs.mkdir(extractDir, { recursive: true });
+  console.log(chalk`{gray Bun v${version}: preparing cache at ${cacheRoot}}`);
+
+  try {
+    await fs.access(zipPath);
+  } catch {
+    console.log(chalk`{gray Bun v${version}: downloading ${url}}`);
+    const res = await fetch(url);
+    if (!res.ok) {
+      return null;
+    }
+    const buf = Buffer.from(await res.arrayBuffer());
+    await fs.mkdir(cacheRoot, { recursive: true });
+    await fs.writeFile(zipPath, buf);
+  }
+
+  try {
+    console.log(
+      chalk`{gray Bun v${version}: extracting archive to ${extractDir}}`,
+    );
+    spawn('unzip', ['-o', zipPath, '-d', extractDir]);
+  } catch (e) {
+    console.warn(
+      chalk`{yellow Bun v${version}: failed to extract (${String(e)})}`,
+    );
+    return null;
+  }
+
+  const bin = await findBunBinaryPath(extractDir);
+  if (!bin) {
+    return null;
+  }
+  console.log(chalk`{gray Bun v${version}: found binary at ${bin}}`);
+  try {
+    await fs.chmod(bin, 0o755);
+  } catch {}
+  return bin;
+};
+
+/**
+ * Gets Bun version info for a given version.
+ * @param version - The Bun version to fetch info for.
+ * @returns An object containing webkitRev and bunRevision, if available.
+ */
+const getBunInfoForVersion = async (
+  version: string,
+): Promise<{ webkitRev?: string; bunRevision?: string }> => {
+  console.log(
+    chalk`{gray Bun v${version}: resolving engine_version and revision...}`,
+  );
+  const bin = await downloadAndExtractBun(version);
+  if (!bin) {
+    return {};
+  }
+  try {
+    const versionsJSON = spawn(bin, [
+      '--eval',
+      'console.log(JSON.stringify(process.versions))',
+    ]);
```

Comment:
I understand that this is currently the only option to get this information, but I'm not comfortable running a downloaded executable in our (privileged) release update workflow. I might have to think about this a bit more.

Is there any chance you could run this in some repo in the oven-sh org? You would probably get something like https://nodejs.org/dist/index.json as a side-effect.

Reply:
Definitely, will set this up now! 🙂